cluster/nginx.go

```diff
@@ -140,18 +117,71 @@ func DeleteNginxLBDeployments(clientset *kubernetes.Clientset, deploymentName st
 		}
 		for _, deployment := range deployments.Items {
 			if deployment.Name != deploymentName {
-				// This delete option is to make sure that the pods belonging to
-				// the nginx-resolver deployment are 'collected' immediately.
 				fgPropagation := metav1.DeletePropagationForeground
 				fgDeleteOption := metav1.DeleteOptions{
 					PropagationPolicy: &fgPropagation,
 				}
-				err := extV1beta1API(clientset).Deployments("default").Delete(deployment.Name, &fgDeleteOption)
-				if err != nil {
-					close(doneCh) // the informer listening for delete deployment event needs to be stopped
-				}
+				extV1beta1API(clientset).Deployments("default").Delete(deployment.Name, &fgDeleteOption)
 			}
 		}
+		fmt.Println("Old nginx-resolver deployments deleted correctly")
+		close(doneCh)
+	}()
+	return doneCh
+}
+
+func CreateNginxResolverDeployment(clientset *kubernetes.Clientset, deploymentName string) <-chan struct{} {
+	doneCh := make(chan struct{})
+	go func() {
+		factory := informers.NewSharedInformerFactory(clientset, 0)
+		deploymentInformer := factory.Extensions().V1beta1().Deployments().Informer()
+		deploymentInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				deployment := obj.(*v1beta1.Deployment)
+				if deployment.GetLabels()["app"] == deploymentName {
+					fmt.Println("nginx-resolver deployment created correctly")
+					close(doneCh)
+				}
+			},
+		})
+
+		go deploymentInformer.Run(doneCh)
+
+		//Creating nginx-resolver deployment with new rules
+		nginxLBDeployment.Spec.Template.ObjectMeta.Labels["app"] = deploymentName
+		nginxLBDeployment.ObjectMeta.Name = deploymentName
+		_, err := extV1beta1API(clientset).Deployments("default").Create(&nginxLBDeployment)
+		if err != nil {
+			close(doneCh)
+		}
+	}()
+	return doneCh
+}
+
+func UpdateNginxResolverService(clientset *kubernetes.Clientset) <-chan struct{} {
+	doneCh := make(chan struct{})
+	go func() {
+		factory := informers.NewSharedInformerFactory(clientset, 0)
+		serviceInformer := factory.Core().V1().Services().Informer()
+		serviceInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+			UpdateFunc: func(oldObj, obj interface{}) {
+				service := obj.(*v1.Service)
+				if service.GetLabels()["name"] == "nginx-resolver" {
```

Comment:
To confirm that the UpdateFunc is seeing the effect of the update from this function, we should check that the updated selector is same as what this function updated it to.